src/home_robot/home_robot/agent/ovmm_agent/llm_agent.py

```diff
@@ -0,0 +1,169 @@
+            if object_id not in obs.semantic:
+                continue
+
+            if (
+                self._get_num_goal_pixels(obs, object_id)
+                < self.num_object_pixel_threshold
+            ):
+                continue
+
+            image_arr = self._segment_goal(obs, object_id)
+            image = (
+                self.preprocess(Image.fromarray(image_arr)).unsqueeze(0).to(self.device)
+            )
+
+            with torch.no_grad():
+                image_features = self.model.encode_image(image)
+                # text_features = self.model.encode_text(text)
+                logits_per_image, logits_per_text = self.model(image, text)
+                probs = logits_per_image.softmax(dim=-1).cpu().numpy()
+                print(probs)
+                if np.max(probs) > self.clip_threshold:
+                    # if True:
+                    self._save_image(image_arr)
+                    print(self.object_names[np.argmax(probs)])
+                    print("Label probs:", probs)
+
+                    self.memory[self.timesteps[0]]["clip_features"].append(
+                        image_features.squeeze(0).tolist()
+                    )
+
+            # self.memory.append(image_features.squeeze(0))
+
+    def _segment_goal(self, obs, object_id):
+        segmented_rgb = np.zeros(obs.rgb.shape, dtype=np.uint8)
+        min_h = obs.rgb.shape[0]
+        min_w = obs.rgb.shape[1]
+        max_h = 0
+        max_w = 0
+
+        for i, ele in np.ndenumerate(obs.semantic):
+            if ele == object_id:
+                segmented_rgb[i] = obs.rgb[i]
+                if i[0] < min_h:
+                    min_h = i[0]
+                if i[1] < min_w:
+                    min_w = i[1]
+                if i[0] > max_h:
+                    max_h = i[0]
+                if i[1] > max_w:
+                    max_w = i[1]
+
+        cropped = np.zeros(
+            (max_h - min_h, max_w - min_w, obs.rgb.shape[2]), dtype=np.uint8
+        )
+        for h in range(cropped.shape[0]):
+            for w in range(cropped.shape[1]):
+                cropped[h, w] = segmented_rgb[h + min_h, w + min_w]
+        return cropped
+
+    #      def _is_object_perceived(self, obs):
+    #      # now using GT segments
+    #      # object_id = obs.task_observations['object_goal']
+    #      for object_id in self.object_of_interest:
+    #          if object_id not in self.goal_obs: self.goal_obs[object_id] = None
+    #          if object_id in obs.semantic:
+    #              num_goal_pixels = self._get_num_goal_pixels(obs, object_id)
+    #              if num_goal_pixels > self._get_num_goal_pixels(
+    #                      self.goal_obs[object_id], object_id):
+    #                  self.goal_obs[object_id] = obs
+    #          elif self.goal_obs[object_id]:
+    #              return True
+    #      return False
+    #
+
+    def _get_num_goal_pixels(self, obs, object_id):
```

Comment:
This is currently visible goal pixels?

Reply:
Yes it is. Added a simple docstring to explain